### ISSUE TITLE
db.collection.find with $in operator and array of ObjectId's

### ISSUE THREAD

**Issue by willtchiu** [mongodb-js/extended-json#23]
I'm trying to query documents in a collection by providing an array of _id's and there's some trouble in parsing into JSON with the $in query operator. The string that I'm parsing is:

```
{
    "_id": {
        "$in": [
            ObjectId("55f2132f789c29a64321ed0d"),
            ObjectId("55f7a34d3a4ffb4b40e4e8e3"),
            ObjectId("55f8d88a3a4ffb4b40e4e8e6")
        ]
    }
}
```

I was able to query successfully in MongoShell using the exact same string as above, however I think it's because ObjectId() is only used for MongoShell mode.

I'm able to query my collection by using $oid with the $or operator:

```
{
    "$or": [
        {
            "_id": {
                "$oid": "55f2132f789c29a64321ed0d"
            }
        },
        {
            "_id": {
                "$oid": "55f7a34d3a4ffb4b40e4e8e3"
            }
        },
        {
            "_id": {
                "$oid": "55f8d88a3a4ffb4b40e4e8e6"
            }
        }
    ]
}
```

I've also tried the strict mode with $oid and $in. I've worked around the problem by using $or instead of $in. I'm wondering if EJSON can parse whenever $in is used in conjunction with an array of ObjectIds.

**Comment by rueckstiess** [comment]
Hi @willtchiu,

I'm not sure I understand what you want to achieve. Can you give a code example how you use `EJSON.parse()` or `EJSON.stringify()` and what you expect it to return?

This example works for me:
```js
var ObjectId = require('bson').ObjectId;
var EJSON = require('mongodb-extended-json');

var doc = { 
  _id: {
    $in: [
      ObjectId('55f2132f789c29a64321ed0d'),
      ObjectId('55f7a34d3a4ffb4b40e4e8e3'),
      ObjectId('55f8d88a3a4ffb4b40e4e8e6')
    ]
  }
};

console.log(EJSON.stringify(doc));
```
It prints out the following string:

```
{"_id":{"$in":[{"$oid":"55f2132f789c29a64321ed0d"},{"$oid":"55f7a34d3a4ffb4b40e4e8e3"},{"$oid":"55f8d88a3a4ffb4b40e4e8e6"}]}}
```

Note that if you have problems with MongoDB directly (i.e. it can't parse your query) that is unrelated to this module. It only converts Javascript objects to EJSON strings and vice versa.

**Comment by willtchiu** [comment]
Hey @rueckstiess,

Thanks for answering. I am constructing a EJSON string in my iOS app and packaging it as a query request to my server. I'm trying to query in a specific collection all documents that match the array of _ids that i've sent through as an request. The query is a string and I'm trying to parse it with EJSON.parse() in order to send it to my query function. 

I have the following string sent through the request:

```
{"_id":{"$in":[ObjectId('55f2132f789c29a64321ed0d'),ObjectId('55f7a34d3a4ffb4b40e4e8e3'),ObjectId('55f8d88a3a4ffb4b40e4e8e6')]}}
```

I get an unexpected token O. What I'm ultimately trying to achieve is to be able to parse an EJSON string with the $in operator which contains an array of ObjectIds, so that I'm able to then call db.collection.find(query).

**Comment by rueckstiess** [comment]
Ah, got it. The problem is in your EJSON string, you use `ObjectId('55f2132f789c29a64321ed0d')`, but that's not EJSON conform. It needs to be `{"$oid": "55f2132f789c29a64321ed0d"}` instead. Hence the `Unexpected token O` error message.

Here an example in the node shell: 

```js
> var EJSON = require('mongodb-extended-json');
> parsed = EJSON.parse('{"_id":{"$in":[{"$oid":"55f2132f789c29a64321ed0d"},{"$oid":"55f7a34d3a4ffb4b40e4e8e3"},{"$oid":"55f8d88a3a4ffb4b40e4e8e6"}]}}')
{ _id:
   { '$in':
      [ 55f2132f789c29a64321ed0d,
        55f7a34d3a4ffb4b40e4e8e3,
        55f8d88a3a4ffb4b40e4e8e6 ] } }

// The hex strings really are ObjectIDs, that's just how they are displayed
> parsed._id.$in[0]._bsontype
'ObjectID'
```

So you'd have to change your code in your iOS app to construct the string for ObjectIDs correctly.

**Comment by willtchiu** [comment]
I had initially used the strict mode format to parse my EJSON strings, however I ran into a problem when I tried to parse:

```
{"_id":{"$in":[{"$oid":"55f2132f789c29a64321ed0d"},{"$oid":"55f7a34d3a4ffb4b40e4e8e3"},{"$oid":"55f8d88a3a4ffb4b40e4e8e6"}]}}
```

After reviewing your latest example, I realized that "$oid":"_id" should be included in braces and that was the reason behind other unexpected token errors I was receiving. I forgot for key/pairs even within an array, that I needed to include that in braces as well.

Thanks a bunch for clarifying this to me!
